src/dmd/must_use.d

```diff
@@ -0,0 +1,198 @@
+    // semantic analysis of `sym` is still in progress
+    // TODO: factor out common code from this function and checkGNUABITag
+    foreach (exp; *sym.userAttribDecl.atts)
+    {
+        if (isMustUseAttribute(exp))
+        {
+            if (sym.isFuncDeclaration())
+            {
+                error(sym.loc, "`@%s` on functions is reserved for future use",
+                    Id.udaMustUse.toChars());
+                sym.errors = true;
+            }
+            else if (sym.isClassDeclaration() || sym.isEnumDeclaration())
+            {
+                error(sym.loc, "`@%s` on `%s` types is reserved for future use",
+                    Id.udaMustUse.toChars(), sym.kind());
+                sym.errors = true;
+            }
+        }
+    }
+}
+
+/**
+ * Returns true if the given expression is an assignment, either simple (a = b)
+ * or compound (a += b, etc).
+ */
+private bool isAssignment(Expression e)
+{
+    import dmd.common.string : startsWith, endsWith;
+
+    if (e.isAssignExp || e.isBinAssignExp)
+        return true;
+    if (auto ce = e.isCallExp())
+    {
+        auto fd = ce.f;
+        auto id = fd ? fd.ident : null;
+        // opXXX are reserved identifiers, so it's ok to match too much here
+        if (id && id.toString().startsWith("op") && id.toString().endsWith("Assign"))
+            return true;
+    }
+    return false;
+}
+
+/**
+ * Returns true if the given expression is an increment (++) or decrement (--).
+ */
+private bool isIncrementOrDecrement(Expression e)
+{
+    import dmd.dtemplate : isExpression;
+    import dmd.globals : Loc;
+    import dmd.id : Id;
+    import dmd.tokens : EXP;
+
+    if (e.op == EXP.plusPlus
+        || e.op == EXP.minusMinus
+        || e.op == EXP.prePlusPlus
+        || e.op == EXP.preMinusMinus)
+        return true;
+    if (auto call = e.isCallExp())
+    {
+        // Check for overloaded preincrement
+        // e.g., a.opUnary!"++"
+        auto fd = call.f;
+        auto id = fd ? fd.ident : null;
+        if (id == Id.opUnary)
+        {
+            auto ti = fd.parent ? fd.parent.isTemplateInstance() : null;
+            auto tiargs = ti ? ti.tiargs : null;
+            if (tiargs && tiargs.length >= 1)
+            {
+                auto argExp = (*tiargs)[0].isExpression();
+                auto op = argExp ? argExp.isStringExp() : null;
+                scope plusPlus = new StringExp(Loc.initial, "++");
+                scope minusMinus = new StringExp(Loc.initial, "--");
+                if (op && (op.compare(plusPlus) == 0 || op.compare(minusMinus) == 0))
+                    return true;
+            }
+        }
+    }
+    else if (auto comma = e.isCommaExp())
+    {
+        // Check for overloaded postincrement
+        // e.g., (auto tmp = a, ++a, tmp)
+        auto left = comma.e1 ? comma.e1.isCommaExp() : null;
+        auto middle = left ? left.e2 : null;
+        if (middle && isIncrementOrDecrement(middle))
```

Comment:
These three lines have multiple redundant tests.

Reply:
I'm afraid it's not clear to me which tests here are redundant. By my count, there are 3, and all test separate pointers:

* `comma.e1`
* `comma.e1.isCommaExp()` (`left`)
* `comma.e1.isCommaExp().e2` (`middle`)

Is there some undocumented invariant that would allow one or more of these tests to be skipped? (E.g., "`e1` and `e2` of a  `CommaExp` are never null".)

Reply:
!comma.e1 is tested which then sets left to null, which is then tested for null to set middle, then middle is tested for null.

Reply:
For example, if comma.e1 is false, none of the rest of the code needs to be executed.